server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/symbols/SymbolAccumulator.java

```diff
@@ -97,62 +98,82 @@ public void registerCodeBlock(ProgramNode program, CodeBlockDefinitionNode node)
+      ref.getUsage().add(usageNode.getLocality().toLocation());
+      if (ref.getDefinitions().size() > 1) {
+        ambiguous = true;
       }
     }
+    return ambiguous
+            ? Optional.of(
+                  SyntaxError.syntaxError()
+                          .errorSource(ErrorSource.PARSING)
+                          .messageTemplate(
+                                  MessageTemplate.of("semantics.ambiguous", usageNode.getName()))
+                          .severity(ErrorSeverity.ERROR)
+                          .location(usageNode.getLocality().toOriginalLocation())
+                          .build())
+            : Optional.empty();
+  }
 
-    CodeBlockDefinitionNode definition = definitions.get(0);
-    definition.addUsage(node.getLocality());
-
-    Optional.ofNullable(symbolTable.getParagraphMap().get(node.getName()))
-        .ifPresent(it -> it.addUsage(node.getLocality().toLocation()));
-    Optional.ofNullable(symbolTable.getSectionMap().get(node.getName()))
-        .ifPresent(it -> it.addUsage(node.getLocality().toLocation()));
+  private List<CodeBlockReference> resolveProcedureId(CodeBlockUsageNode usageNode, SymbolTable symbolTable) {
+    Map<ProcedureId, CodeBlockReference> procedures = symbolTable.getProcedures();
+    String ofSection = ofSection(usageNode);
+    ProcedureId paragraphId = new ProcedureId(ofSection, usageNode.getName());
+    if (procedures.containsKey(paragraphId)) {
+      return Collections.singletonList(procedures.get(paragraphId));
+    }
+    if (ofSection != null) {
+        return Collections.emptyList();
+    }
+    String section = usageNode.getNearestParentByType(NodeType.PROCEDURE_SECTION)
+            .map(n -> ((ProcedureSectionNode) n).getName()).orElse(null);
+    paragraphId = new ProcedureId(section, usageNode.getName());
+    if (procedures.containsKey(paragraphId)) {
+      return Collections.singletonList(procedures.get(paragraphId));
+    }
+    // try section name
+    ProcedureId sectionId = new ProcedureId(usageNode.getName(), null);
+    if (procedures.containsKey(sectionId)) {
+      return Collections.singletonList(procedures.get(sectionId));
+    }
+    List<CodeBlockReference> resolved = new ArrayList<>();
+    for (Map.Entry<ProcedureId, CodeBlockReference> en: symbolTable.getProcedures().entrySet()) {
+      if (Objects.equals(usageNode.getName(), en.getKey().getParagraphName())) {
+        resolved.add(en.getValue());
+      }
+    }
+    return resolved;
+  }
 
-    return Optional.empty();
+  private String ofSection(CodeBlockUsageNode usageNode) {
+      int nextIndex = usageNode.getParent().getChildren().indexOf(usageNode) + 1;
+      // Chech if there is OF/IN section name
+      if (nextIndex < usageNode.getParent().getChildren().size()) {
+        Node c = usageNode.getParent().getChildren().get(nextIndex);
+        if (c instanceof SectionNameNode) {
+          return ((SectionNameNode) c).getName();
+        }
+      }
+      return null;
```

Comment:
Wouldn't it be easier to replace 
https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/blob/development/server/engine/src/main/java/org/eclipse/lsp/cobol/core/visitor/CobolVisitor.java#L940-L943
with something like
```
  @Override
  public List<Node> visitProcedureName(ProcedureNameContext ctx) {
    final ParagraphNameContext pCtx = ctx.paragraphName();
    final InSectionContext isCtx = ctx.inSection();
    final SectionNameContext sCtx = isCtx == null ? null : isCtx.sectionName();
    return addTreeNode(ctx, locality -> new CodeBlockUsageNode(
          locality,
          VisitorHelper.getName(pCtx),
          sCtx != null ? VisitorHelper.getName(sCtx) : null));
  }

```
after adding nullable `sectionName` into `CodeBlockUsageNode`?